### DISCUSSION TITLE
What is the correct way to translate ordinal numbers?

### DISCUSSION THREAD

**Discussion by mrtnzlml** [facebookincubator/fbt#307] [Q&A] [answered]
With plural (cardinal) numbers it's really simple: to my best knowledge, you mark the plurals with `fbt:plural`, extract all combinations (which is 2 for English), and via variations, you can translate any number of plurals for the specific language.

So say in English ([CLDR](https://unicode-org.github.io/cldr-staging/charts/40/supplemental/language_plural_rules.html#en)), you'd extract:

```text
ONE:   {number} car
OTHER: {number} cars
```

But in Czech ([CLDR](https://unicode-org.github.io/cldr-staging/charts/40/supplemental/language_plural_rules.html#cs)), the translator would prepare 3 variants that you can apply easily since FBT understand Czech plurals thanks to CLDR:

```text
ONE:   {number} auto
FEW:   {number} auta
OTHER: {number} aut
```

For Arabic, you'd prepare 6 variants. Easy enough.

Now, the ordinal numbers are a bit different. In English you'd need:

```text
ONE:   1st car, 21st car, …
TWO:   2nd car, 22nd car, …
FEW:   3rd car, 23rd car, …
OTHER: 4th car, 11th car, …
```

These 4 combinations have to translate into only one ordinal variant in Czech though:

```text
OTHER: 1. auto
```

So the question is: what is the best way how to handle ordinal numbers in FBT? 🤔 

Thank you very much for your insights! 😎

**Comment by pkqinys** [comment, marked as answer]
Unfortunately, akaik there isn't an easy way to write numbers in ordinal form. One workaround is to use `fbt:enum` and write all ordinals explicitly as enum values. 

I.e. 
```
  FIRST: 1st car
  SECOND: 2nd car
  THIRD: 3rd car
  FOURTH: 4th car
  ...
  ELEVENTH: 11th car
  ...
  TWENTY_FIRST: 21st car
  TWENTY_SECOND: 22nd car
  TWENTY_THIRD: 23rd car
```

**Reply by mrtnzlml** [reply]
I see, so something like this, right?

```js
<fbt desc="example string">
  <fbt:enum
    enum-range={{
      FIRST: '1st car',
      SECOND: '2nd car',
      THIRD: '3rd car',
      FOURTH: '4th car',
    }}
    value={'FIRST'}
  />{' '}
  wins!
</fbt>
```

I guess that's good enough for simple situations. For more complex cases you can always downgrade to simple JS logic:

```js
const getOrdinalString = (n) => {
  if (n % 10 === 1 && n % 100 !== 11) {
    return fbt(fbt.param('number', n) + 'st car wins!', 'example');
  } else if (n % 10 === 2 && n % 100 !== 12) {
    return fbt(fbt.param('number', n) + 'nd car wins!', 'example');
  } else if (n % 10 === 3 && n % 100 !== 13) {
    return fbt(fbt.param('number', n) + 'rd car wins!', 'example');
  } else {
    return fbt(fbt.param('number', n) + 'th car wins!', 'example');
  }
};
```

Thank you for your answer! 😎
